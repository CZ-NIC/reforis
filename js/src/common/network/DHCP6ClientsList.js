/*
 * Copyright (C) 2019-2021 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import PropTypes from "prop-types";
import moment from "moment";
import { formFieldsSize } from "foris";

DHCP6ClientsList.propTypes = {
    ipv6clients: PropTypes.arrayOf(PropTypes.object),
};

export default function DHCP6ClientsList({ ipv6clients }) {
    return (
        <div className={formFieldsSize}>
            <h2>{_("IPv6 DHCP Client List")}</h2>
            <p>
                {_(
                    "This list contains all devices that are connected to the network through wired or wireless connections using IPv6."
                )}
            </p>
            {ipv6clients.length === 0 ? (
                <p className="text-muted text-center">
                    {_("No clients found.")}
                </p>
            ) : (
                <div className="table-responsive">
                    <table className="table table-hover">
                        <thead className="thead-light">
                            <tr className="text-left">
                                <th>{_("Hostname")}</th>
                                <th>{_("IPv6 Address")}</th>
                                <th>{_("DUID")}</th>
                                <th className="text-center">{_("Expires")}</th>
                            </tr>
                        </thead>
                        <tbody>
                            {ipv6clients.map((client) => (
                                <DHCP6ClientsListItem
                                    key={client.ipv6}
                                    {...client}
                                />
                            ))}
                        </tbody>
                    </table>
                </div>
            )}
        </div>
    );
}

DHCP6ClientsListItem.propTypes = {
    ipv6: PropTypes.string.isRequired,
    expires: PropTypes.number.isRequired,
    duid: PropTypes.string.isRequired,
    hostname: PropTypes.string.isRequired,
};

function DHCP6ClientsListItem({ ipv6, expires, duid, hostname }) {
    return (
        <tr className="text-left">
            <td>{hostname}</td>
            <td>{ipv6}</td>
            <td>{duid}</td>
            <td className="text-center">
                {moment.unix(expires).format("YYYY-MM-DD HH:mm")}
            </td>
        </tr>
    );
}
