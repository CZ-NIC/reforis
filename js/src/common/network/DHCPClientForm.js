/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from 'react';
import propTypes from 'prop-types';

import TextInput from '../bootstrap/TextInput';
import {validateDomain} from '../validations';

const HELP_TEXTS = {
    hostname: _('Hostname which will be provided to DHCP server.'),
};

DHCPClientForm.propTypes = {
    formData: propTypes.shape({
        hostname: propTypes.string,
    }).isRequired,
    formErrors: propTypes.shape({
        hostname: propTypes.string
    }),
    setFormValue: propTypes.func.isRequired,
    updateRule: propTypes.func.isRequired,
};

export default function DHCPClientForm({formData, formErrors, setFormValue, updateRule, ...props}) {
    return <TextInput
        label={_('DHCP hostname')}
        value={formData.hostname || ''}
        error={(formErrors || {}).hostname || null}
        helpText={HELP_TEXTS.hostname}

        onChange={setFormValue(
            value => updateRule({hostname: {$set: value}})
        )}

        {...props}
    />
}

export function validateDHCPForm(formData) {
    const error = {hostname: validateDomain(formData.hostname)};
    return error.hostname ? error : null;
}
