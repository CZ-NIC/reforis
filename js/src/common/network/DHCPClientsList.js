/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import PropTypes from "prop-types";
import moment from "moment";

DHCPClientsList.propTypes = {
    clients: PropTypes.arrayOf(PropTypes.object),
};

export default function DHCPClientsList({ clients }) {
    return (
            <h2>{_("DHCP Client List")}</h2>
            <p>
                {_(
                    "This list contains all devices that are connected to the network through wired or wireless connections."
                )}
            </p>
            {clients.length === 0 ? (
                <p className="text-muted text-center">
                    {_("No DHCP clients found.")}
                </p>
            ) : (
                <div className="table-responsive">
                    <table className="table table-hover">
                        <thead>
                            <tr className="text-center">
                                <th>{_("Expires")}</th>
                                <th>{_("IP Address")}</th>
                                <th>{_("MAC Address")}</th>
                                <th>{_("Hostname")}</th>
                                <th>{_("Active")}</th>
                            </tr>
                        </thead>
                        <tbody>
                            {clients.map((client) => (
                                <DHCPClientsListItem
                                    key={client.ip}
                                    {...client}
                                />
                            ))}
                        </tbody>
                    </table>
                </div>
            )}
        </>
    );
}

DHCPClientsListItem.propTypes = {
    ip: PropTypes.string.isRequired,
    expires: PropTypes.number.isRequired,
    mac: PropTypes.string.isRequired,
    hostname: PropTypes.string.isRequired,
    active: PropTypes.bool.isRequired,
};

function DHCPClientsListItem({ ip, expires, mac, hostname, active }) {
    return (
        <tr className="text-center">
            <td>{moment.unix(expires).format("YYYY-MM-DD HH:mm")}</td>
            <td>{ip}</td>
            <td>{mac}</td>
            <td>{hostname}</td>
            <td>
                <i className={`fas ${active ? "fa-check" : "fa-times"}`} />
            </td>
        </tr>
    );
}
