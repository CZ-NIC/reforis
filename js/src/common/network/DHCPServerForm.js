/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from 'react';
import propTypes from 'prop-types';

import NumberInput from 'common/bootstrap/NumberInput';

export const HELP_TEXT = _('Enable this option to automatically assign IP addresses to the devices connected to the router.');

DHCPServerForm.propTypes = {
    formData: propTypes.shape({
        start: propTypes.number,
        limit: propTypes.number,
        lease_time: propTypes.number,
    }).isRequired,
    setFormValue: propTypes.func.isRequired,
    updateRule: propTypes.func.isRequired,
};

export default function DHCPServerForm({formData, updateRule, setFormValue, ...props}) {
    return <>
        <NumberInput
            label={_('DHCP start')}
            value={formData.start}
            min='1'
            required

            onChange={setFormValue(
                value => updateRule({start: {$set: value}})
            )}

            {...props}
        />
        <NumberInput
            label={_('DHCP max leases')}
            value={formData.limit}
            min='1'
            required

            onChange={setFormValue(
                value => updateRule({limit: {$set: value}})
            )}

            {...props}
        />
        <NumberInput
            label={_('Lease time (hours)')}
            value={formData.lease_time}
            min='120'
            required

            onChange={setFormValue(
                value => updateRule({lease_time: {$set: value}})
            )}

            {...props}
        />
    </>
}
