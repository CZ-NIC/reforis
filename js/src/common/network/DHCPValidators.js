/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import ipaddr from "ipaddr";

import { undefinedIfEmpty, withoutUndefinedKeys, validateIPv4Address } from "foris";

import { validateRequiredField } from "./validators";
import { addToAddress } from "./utils";

/*
 * Check if DHCP start address is within network.
 * - networkAddress - address object (parsed by ipaddr)
 * - prefixLength - network prefix related to mask (e.g. 24 for 255.255.255.0)
 * - DHCPStartAddress - address object (parsed by ipaddr)
 */
function isDHCPStartWithinNetwork(networkAddress, prefixLength, DHCPStartAddress) {
    // Check if DHCP start address is within network
    return DHCPStartAddress.match(networkAddress, prefixLength);
}

/*
 * Check if DHCP start address clashes with a reserved address.
 * - networkAddress - address object (parsed by ipaddr)
 * - broadcastAddress - address object (parsed by ipaddr)
 * - DHCPStartAddress - address object (parsed by ipaddr)
 */
function isDHCPStartReserved(networkAddress, broadcastAddress, DHCPStartAddress) {
    // Check if DHCP start address is not at the borders of address range
    if ([broadcastAddress.toString(), networkAddress.toString()]
        .includes(DHCPStartAddress.toString())
    ) {
        return true;
    }
    return false;
}

/*
 * Check if number of leases can be realized in this network.
 * - networkAddress - address object (parsed by ipaddr)
 * - prefixLength - address object (parsed by ipaddr)
 * - DHCPStart - address object (parsed by ipaddr)
 * - maxLeases - maximum number of addresses used by DHCP
 */
function isMaxLeasesValid(networkAddress, prefixLength, DHCPStart, maxLeases) {
    const maxAddress = addToAddress(DHCPStart, maxLeases - 1);
    return maxAddress.match(networkAddress, prefixLength);
}

function validatePositiveNumber(number) {
    // eslint-disable-next-line no-restricted-globals
    if (number === "" || isNaN(number)) {
        return _("Value must be a number.");
    } if (number < 1) {
        return _("Value must be positive.");
    }
    return undefined;
}

export default function validateDHCP(
    dhcpSettings, routerIPString, networkMask, isNetworkConfigBroken,
) {
    if (isNetworkConfigBroken) {
        return {
            start: _("Invalid network settings. DHCP settings can't be validated."),
        };
    }

    const errors = {
        start: (
            validateRequiredField(dhcpSettings.start)
            || validateIPv4Address(dhcpSettings.start)
        ),
        limit: validatePositiveNumber(dhcpSettings.limit),
        lease_time: dhcpSettings.lease_time < 1 ? _("Minimum lease time is 1 hour.") : undefined,
    };

    if (!errors.start) {
        const prefixLength = ipaddr.IPv4.parse(networkMask).prefixLengthFromSubnetMask();
        const CIDR = `${routerIPString}/${prefixLength}`;
        const networkAddress = ipaddr.IPv4.networkAddressFromCIDR(CIDR);
        const broadcastAddress = ipaddr.IPv4.broadcastAddressFromCIDR(CIDR);
        const DHCPStart = ipaddr.IPv4.parse(dhcpSettings.start);

        if (!isDHCPStartWithinNetwork(networkAddress, prefixLength, DHCPStart)) {
            errors.start = _("Address is outside current network.");
        } else if (isDHCPStartReserved(networkAddress, broadcastAddress, DHCPStart)) {
            errors.start = _("Address is already reserved for other purposes.");
        } else if (!errors.limit) { // There are no more important errors so far.
            if (!isMaxLeasesValid(networkAddress, prefixLength, DHCPStart, dhcpSettings.limit)) {
                errors.limit = _("Too many addresses requested. Set a lower number or change DHCP start.");
            } else if (DHCPStart.toString() === routerIPString && dhcpSettings.limit === 1) {
                errors.limit = _("The only DHCP lease is the same as router's address. Increase limit or change start address.");
            }
        }
    }

    return undefinedIfEmpty(withoutUndefinedKeys(errors));
}
