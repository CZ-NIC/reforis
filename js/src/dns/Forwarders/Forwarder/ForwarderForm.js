/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import PropTypes from "prop-types";
import {
    Button, Select, Spinner, TextInput, API_STATE,
} from "foris";

import forwarderPropTypes from "../propTypes";
import useForwarderForm from "./hooks";
import IPAddressesForm from "./IPAddressesForm";

const TLS_TYPES = {
    no: "No TLS",
    hostname: "Hostname",
    pin: "Pin",
};

const HELP_TEXTS = {
    tls_pin: _("base64 encoded sha256."),
};

ForwarderForm.propTypes = {
    forwarder: forwarderPropTypes,
    saveForwarderCallback: PropTypes.func,
};

export default function ForwarderForm({ forwarder, saveForwarderCallback }) {
    const [
        formState,
        setFormValue,
        postState,
        saveForwarder,
    ] = useForwarderForm(forwarder, saveForwarderCallback);

    if (!formState.data) return <Spinner className="text-center" />;

    const disabled = postState.state === API_STATE.SENDING;
    const formData = formState.data;
    const formErrors = formState.errors || {};

    return (
        <>
            <TextInput
                label={_("Name")}
                value={formData.description}
                error={formErrors.description}

                onChange={setFormValue(
                    (value) => ({ description: { $set: value } }),
                )}
                disabled={disabled}
            />
            <IPAddressesForm
                ipVersion="ipv4"
                ipaddresses={formData.ipaddresses.ipv4}
                setFormValue={setFormValue}
                errors={(formErrors.ipaddresses || {}).ipv4}
                disabled={disabled}
            />
            <IPAddressesForm
                ipVersion="ipv6"
                ipaddresses={formData.ipaddresses.ipv6}
                setFormValue={setFormValue}
                errors={(formErrors.ipaddresses || {}).ipv6}
                disabled={disabled}
            />
            <Select
                label={_("TLS type")}
                value={formData.tls_type}
                choices={TLS_TYPES}

                onChange={setFormValue(
                    (value) => {
                        const fieldsToAdd = {
                            hostname: "tls_hostname",
                            pin: "tls_pin",
                        };
                        return {
                            tls_type: { $set: value },
                            [fieldsToAdd[value]]: { $set: "" },
                        };
                    },
                )}

                disabled={disabled}
            />
            {formData.tls_type === "hostname" && (
                <TextInput
                    label={_("TLS hostname")}
                    value={formData.tls_hostname || ""}
                    error={formErrors.tls_hostname}

                    onChange={setFormValue(
                        (value) => ({ tls_hostname: { $set: value } }),
                    )}
                    disabled={disabled}
                />
            )}

            {formData.tls_type === "pin" && (
                <TextInput
                    label={_("TLS pin")}
                    value={formData.tls_pin || ""}
                    error={formErrors.tls_pin}
                    helpText={HELP_TEXTS.tls_pin}

                    onChange={setFormValue(
                        (value) => ({ tls_pin: { $set: value } }),
                    )}
                    disabled={disabled}
                />
            )}
            <Button
                onClick={saveForwarder}
                forisFormSize
                disabled={!!formState.errors}
            >
                {_("Save forwarder")}
            </Button>
        </>
    );
}
