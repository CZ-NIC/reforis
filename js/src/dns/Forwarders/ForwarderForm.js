/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import PropTypes from "prop-types";
import { Select, TextInput } from "foris";

import forwarderPropTypes from "./propTypes";

const TLS_TYPES = {
    no: "No TLS",
    hostname: "Hostname",
    pin: "Pin",
};

const HELP_TEXTS = {
    tls_pin: _("base64 encoded sha256."),
};

ForwarderForm.propTypes = {
    formData: forwarderPropTypes,
    formErrors: PropTypes.shape({
        ipaddresses: PropTypes.shape({
            ipv4: PropTypes.string,
            ipv6: PropTypes.string,
        }),
        tls_hostname: PropTypes.string,
    }),
    setFormValue: PropTypes.func,
};

export default function ForwarderForm({
    formData, formErrors, setFormValue, ...props
}) {
    const errors = formErrors || {};
    return (
        <>
            <TextInput
                label={_("Name")}
                value={formData.name}
                error={errors.name}
                required

                onChange={setFormValue(
                    (value) => ({ name: { $set: value } }),
                )}
                {...props}
            />
            <TextInput
                label={_("Description")}
                value={formData.description}

                onChange={setFormValue(
                    (value) => ({ description: { $set: value } }),
                )}
                {...props}
            />
            <TextInput
                label={_("IPv4")}
                value={formData.ipaddresses.ipv4}
                error={(errors.ipaddresses || {}).ipv4}

                onChange={setFormValue(
                    (value) => ({ ipaddresses: { ipv4: { $set: value } } }),
                )}
                {...props}
            />
            <TextInput
                label={_("IPv6")}
                value={formData.ipaddresses.ipv6}

                error={(errors.ipaddresses || {}).ipv6}
                onChange={setFormValue(
                    (value) => ({ ipaddresses: { ipv6: { $set: value } } }),
                )}
                {...props}
            />
            <Select
                label={_("TLS type")}
                value={formData.tls_type}
                choices={TLS_TYPES}

                onChange={setFormValue(
                    (value) => {
                        const fieldsToAdd = {
                            hostname: "tls_hostname",
                            pin: "tls_pin",
                        };
                        return {
                            tls_type: { $set: value },
                            [fieldsToAdd[value]]: { $set: "" },
                        };
                    },
                )}

                {...props}
            />
            {formData.tls_type === "hostname"
                ? (
                    <TextInput
                        label={_("TLS hostname")}
                        value={formData.tls_hostname || ""}
                        error={errors.tls_hostname}

                        onChange={setFormValue(
                            (value) => ({ tls_hostname: { $set: value } }),
                        )}
                        {...props}
                    />
                )
                : null}

            {formData.tls_type === "pin"
                ? (
                    <TextInput
                        label={_("TLS pin")}
                        value={formData.tls_pin || ""}
                        error={errors.tls_pin}
                        helpText={HELP_TEXTS.tls_pin}

                        onChange={setFormValue(
                            (value) => ({ tls_pin: { $set: value } }),
                        )}
                        {...props}
                    />
                )
                : null}
        </>
    );
}
