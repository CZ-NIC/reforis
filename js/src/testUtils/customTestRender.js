/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

/* eslint import/export: "off" */

import React from 'react';
import {UIDReset} from 'react-uid';
import {StaticRouter} from 'react-router';
import {render} from '@testing-library/react'

function Wrapper({children}) {
    return <StaticRouter>
        <UIDReset>
            {children}
        </UIDReset>
    </StaticRouter>
}

const customTestRender = (ui, options) => render(ui, {wrapper: Wrapper, ...options});

// re-export everything
export * from '@testing-library/react'

// override render method
export {customTestRender as render}
