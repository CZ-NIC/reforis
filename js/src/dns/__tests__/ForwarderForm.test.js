/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import mockAxios from "jest-mock-axios";
import userEvent from "@testing-library/user-event";
import { fireEvent, render, wait } from "foris/testUtils/customTestRender";
import { mockSetAlert } from "foris/testUtils/alertContextMock";
import { mockJSONError } from "foris/testUtils/network";
import { ALERT_TYPES } from "foris";

import ForwarderForm from "../Forwarders/Forwarder/ForwarderForm";
import { forwardersFixture } from "./__fixtures__/forwarders";

describe("<ForwarderForm/>: new forwarder.", () => {
    let getByText;
    let getByLabelText;
    let container;
    let saveForwarderCallback;

    beforeEach(() => {
        saveForwarderCallback = jest.fn();
        const renderRes = render(<ForwarderForm saveForwarderCallback={saveForwarderCallback} />);
        getByText = renderRes.getByText;
        getByLabelText = renderRes.getByLabelText;
        container = renderRes.container;
    });

    it("Test with snapshot.", () => {
        expect(container)
            .toMatchSnapshot();
    });

    it("Test add new forwarder - handle success.", async () => {
        userEvent.type(getByLabelText("Name"), "Custom forwarder");
        userEvent.type(getByLabelText("IPv4"), "1.2.3.4");
        fireEvent.click(getByText(/Save forwarder/));

        expect(mockAxios.post)
            .toBeCalled();
        const data = {
            description: "Custom forwarder",
            ipaddresses: {
                ipv4: "1.2.3.4",
                ipv6: "",
            },
            tls_type: "no",
        };
        expect(mockAxios.post)
            .toHaveBeenCalledWith("/reforis/api/dns/forwarders", data, expect.anything());
        mockAxios.mockResponse({ data: {} });

        // Handle success
        await wait(() => {
            expect(saveForwarderCallback).toBeCalled();
        });
        expect(mockSetAlert).toBeCalledWith("Forwarder saved successfully.", ALERT_TYPES.SUCCESS);
    });

    it("Test add new forwarder - handle error.", async () => {
        userEvent.type(getByLabelText("Name"), "Custom forwarder");
        userEvent.type(getByLabelText("IPv4"), "1.2.3.4");
        fireEvent.click(getByText(/Save forwarder/));
        mockJSONError();
        await wait(() => {
            expect(mockSetAlert).toBeCalledWith("Can't save forwarder.");
        });
    });
});

describe("<ForwarderForm/>: existed forwarder.", () => {
    let getByText;
    let container;
    let saveForwarderCallback;

    beforeEach(() => {
        saveForwarderCallback = jest.fn();
        const renderRes = render(
            <ForwarderForm
                forwarder={forwardersFixture.forwarders[0]}
                saveForwarderCallback={saveForwarderCallback}
            />,
        );
        getByText = renderRes.getByText;
        container = renderRes.container;
    });

    it("Test with snapshot.", () => {
        expect(container)
            .toMatchSnapshot();
    });

    it("Test add new forwarder - handle success..", async () => {
        fireEvent.click(getByText(/Save forwarder/));

        expect(mockAxios.put)
            .toBeCalled();
        const data = {
            description: "Google",
            ipaddresses: {
                ipv4: "8.8.8.8",
                ipv6: "2001:4860:4860::8888",
            },
            tls_type: "no",
        };
        expect(mockAxios.put)
            .toHaveBeenCalledWith("/reforis/api/dns/forwarders/99_google", data, expect.anything());
        mockAxios.mockResponse({ data: {} });

        // Handle success
        await wait(() => {
            expect(saveForwarderCallback).toBeCalled();
        });
        expect(mockSetAlert).toBeCalledWith("Forwarder added successfully.", ALERT_TYPES.SUCCESS);
    });

    it("Test add new forwarder - handle error.", async () => {
        fireEvent.click(getByText(/Save forwarder/));
        mockJSONError();
        await wait(() => {
            expect(mockSetAlert).toBeCalledWith("Can't add new forwarder.");
        });
    });
});
