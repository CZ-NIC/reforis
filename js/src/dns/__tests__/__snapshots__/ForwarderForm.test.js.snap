// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ForwarderForm/>: existed forwarder. Test with snapshot. 1`] = `
<div>
  <div
    class="form-group"
  >
    <label
      for="1"
    >
      Name
    </label>
    <div
      class="input-group"
    >
      <input
        class="form-control"
        id="1"
        type="text"
        value="Google"
      />
    </div>
  </div>
  <div
    class="form-group"
  >
    <label
      for="2"
    >
      IPv4
    </label>
    <div
      class="input-group"
    >
      <input
        class="form-control"
        id="2"
        type="text"
        value="8.8.8.8"
      />
    </div>
  </div>
  <div
    class="form-group"
  >
    <label
      for="3"
    >
      IPv6
    </label>
    <div
      class="input-group"
    >
      <input
        class="form-control"
        id="3"
        type="text"
        value="2001:4860:4860::8888"
      />
    </div>
  </div>
  <div
    class="form-group"
  >
    <label
      for="4"
    >
      TLS type
    </label>
    <select
      class="custom-select"
      id="4"
    >
      <option
        value="no"
      >
        No TLS
      </option>
      <option
        value="hostname"
      >
        Hostname
      </option>
      <option
        value="pin"
      >
        Pin
      </option>
    </select>
  </div>
  <button
    class="btn btn-primary  col-sm-12 col-lg-3"
    type="button"
  >
     
     
    Save forwarder
  </button>
</div>
`;

exports[`<ForwarderForm/>: new forwarder. Test with snapshot. 1`] = `
<div>
  <div
    class="form-group"
  >
    <label
      for="1"
    >
      Name
    </label>
    <div
      class="input-group"
    >
      <input
        class="form-control  is-invalid"
        id="1"
        type="text"
        value=""
      />
    </div>
    <div
      class="invalid-feedback"
    >
      Forwarder name should have at least one symbol.
    </div>
  </div>
  <div
    class="form-group"
  >
    <label
      for="2"
    >
      IPv4
    </label>
    <div
      class="input-group"
    >
      <input
        class="form-control  is-invalid"
        id="2"
        type="text"
        value=""
      />
    </div>
    <div
      class="invalid-feedback"
    >
      Define at least one of the IP addresses of the forwarder.
    </div>
  </div>
  <div
    class="form-group"
  >
    <label
      for="3"
    >
      IPv6
    </label>
    <div
      class="input-group"
    >
      <input
        class="form-control  is-invalid"
        id="3"
        type="text"
        value=""
      />
    </div>
    <div
      class="invalid-feedback"
    >
      Define at least one of the IP addresses of the forwarder.
    </div>
  </div>
  <div
    class="form-group"
  >
    <label
      for="4"
    >
      TLS type
    </label>
    <select
      class="custom-select"
      id="4"
    >
      <option
        value="no"
      >
        No TLS
      </option>
      <option
        value="hostname"
      >
        Hostname
      </option>
      <option
        value="pin"
      >
        Pin
      </option>
    </select>
  </div>
  <button
    class="btn btn-primary  col-sm-12 col-lg-3"
    disabled=""
    type="button"
  >
     
     
    Save forwarder
  </button>
</div>
`;
